Nanotransen Legal Liability Initiative - Weapon Stats but more RP

SpaceDragon00 opened the pull request:
Also Known As: I made some changes to the initial PR to better comply with criticisms. I'm not exactly sure whether or not Im allowed to PR the changes to a request that was closed, since I couldn't find any examples on the contributing guide, but please feel free to close and let me know if I'm doing something shitty

## About The Pull Request

Following a series of workplace ""Accidents,"" Nanotransen's team of legal experts have decided to outfit a number of HIGHLY DANGEROUS objects with holographic warning labels! Now if you try to claim that pesky workman's comp, you'll be held liable for daring to be near an object marked as a hazard

- Gives an optional warning label hyperlink when you examine a dangerous item
- Warning labels convey the primary offensive statistics of a given weapon in an esoteric manner (Damage, throw damage, armor pen, block chance and more)
- No more raw damage numbers! (Now fully RP!)
- Inspecting ammo gives you information on the projectile loaded inside (And none if you shot it!)
- Inspecting ballistic weapons lets you know the caliber of the magazine (And none if you lost it!)
- Inspecting energy weapons gives you their various modes

## Why It's Good For The Game

I'll be a bit more thorough here to better explain why I think this is a good thing, as well as addressing some comments I got on the last PR

Why I like it: One of the main frustrations I have that lead to the creation of this mechanic is the fact that it's difficult to nigh impossible to tell the basic weapon statistics of quite a few things without looking in the code. As far as I am aware there is no way to know some of the more esoteric weapons statistics (Armour pen, block chance) without looking in the code. This is, in my books, terrible for immersion as the base weapon statistics are very important for gameplay purposes, but looking at the code directly for this information is more game-ifying than anything

Potential Questions:
"What about the 'force' value on hover over?"
That does not provide all of the information this provides. Yes this provides force, but it also provides caliber, throwing force, block chance, armor penetration, and bullet statistics

"Isn't it kind of silly that you can make a weapon and a tag will automatically be applied?"
Not anymore, with the power of holograms! Nanotransen nanny AIs will now automatically detect any dangerous object you happen to be looking at and apply the tag digitally for the sake of your safety!

"Could you make this a function of another object like an analyzer, or part of a hud?"
I _could_ but I feel that unnecessarily complicates the process of ascertaining important information. If there is a high enough demand for it, I could change it, or combine this PR and that [other one](https://github.com/tgstation/tgstation/pull/58858)

"Won't this clog the examine text?"
I guess, technically, but it's only one line since all the information is available through an optional hyperlink. I consider this to be at least as important as some other examine variables that take up a line or more

"How do you hit someone .1 times?"
When you're a hyper-advanced, borderline-parodical corporate hive mind in the far future, you find a way. 

"Why not just use whole numbers for the number of attacks?"
So that miniscule changes in force (Like turning a laser scalpel 'on') can still be properly conveyed

## Changelog
:cl:
qol: Inspecting a weapon shows a warning label. By default, this shows force, throwforce, armour_penetration, and block_chance in an esoteric manner. Inspecting ammo gives information on the shot it produces, inspecting ballistic weapons lets you know their caliber, and inspecting energy weapons gives you their modes and the damage of those modes
/:cl:

## Images

![image](https://user-images.githubusercontent.com/83368538/116833439-2c791700-ab87-11eb-92c6-df9d318e97d0.png)

![image](https://user-images.githubusercontent.com/83368538/116833444-2f740780-ab87-11eb-9007-154d4b85d9c6.png)

![image](https://user-images.githubusercontent.com/83368538/116833447-326ef800-ab87-11eb-891f-e846f9706ca3.png)

SpaceDragon00 committed:
Sing -> Singe Replacement

Self explanatory. Replaces sing with singe

SpaceDragon00 committed:
Removes Scorch, Just In Case

SpaceDragon00 committed:
Weapon Descriptions (And Also Reversion)

Adds support for weapon statistics to be shown as part of examining an item, similar to the tags found on armor. Also, reverts the small changes I made on master because I'm a fucking idiot and made those changes on master.

SpaceDragon00 committed:
Fixing Minor Formatting Mistakes

Making myself look busy

SpaceDragon00 committed:
(Not Yet Functional) Changing Flavor, Minor Code Refactoring

Starting to put shit back into examine()

TODO:
Make it actually work
Merge get_notes and Topic into one proc
Fix the rest of the flavor

SpaceDragon00 committed:
Refactoring, Reflavor, and Improving Weapon Notes

- No longer unnecessarily creates objects to find weapon variables
- All weapon_notes procs now activate as part of clicking the examine hyperlink
- Now outputs the readout as a joined list
- Improves the flavor of the tag in order to be more believable in-universe
- override_notes is now True/False

SpaceDragon00 committed:
No More Dividing by 0 + Format Fixes

- No longer divides by 0 on energy gun modes without damage
- No longer shows armor pen/block chance for
- Actually comments out the weapon tag val code
- Minor description change for ammunition to better specify that we are talking about stamina damage

SpaceDragon00 committed:
Fixing the rounding on stun batons

Changed the rounding from 1 to 0.1 so there is an actual, calculable difference between the display values of tele batons and stun batons

itseasytosee commented:
This is too much and raises too many questions.

wesoda25 commented:
This is awesome but I feel it's a bit too much for normal examine. I think if you added some sort of item/added this functionality to an item it would be great.

YakumoChen commented:
Funny as fuck but you should cut down on the extraneous text. Make certain phrases not appear if they don't go above a nominal threshold, especially for block rate and armour piercing, and throw force if it's not +=8 or something

SpaceDragon00 commented:
I appreciate the feedback so far, I'll let people leave their suggestions and then figure out what to do from there

The reason the description is so 'extra' is because the first time I made this feature is was criticized for being too game-y, so I tried my best to make it make sense within the universe and add flavor to it

The main reason I dont want to associate it with an item is because it makes something designed to be helpful infinitely less so by adding another hoop to go through for information, but if it's what must be done, I'll consider it

SpaceDragon00 commented:
> 
> 
> Funny as fuck but you should cut down on the extraneous text. Make certain phrases not appear if they don't go above a nominal threshold, especially for block rate and armour piercing, and throw force if it's not +=8 or something
> 
> Also you should really use rounding.

Armor pen and block chance only show up if either values aren't 0, and the warning doesn't appear aswell if the force and throwforce are too low (4 or less). The miscommunication is my bad because I took the Double-Barrel screenshot before I made the change

Throwforce still shows up as normal when the actual notes are called primarily because it lets people know explicitly that the item does nothing when thrown, however I can cut it out if need be

YakumoChen commented on code/game/objects/items.dm:
```diff
@@ -337,6 +346,82 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+ *
+ */
+/obj/item/proc/warning_label(var/list/readout = "")
+	var/list/crimes = list("Assaults", "Third Degree Murders", "Robberies", "Terrorist Attacks", "Different Felonies", "Felinies", "Counts of Tax Evasion", "Mutinies")
```
Unauthorized Slapstick Comedies
Nonviolent Protests
Assassinations
[REDACTED]
Shipjackings

cacogen replied:
This stuff would look better if it was lowercase and you misspelled felonies once

MrDoomBringer replied:
feline

SpaceDragon00 replied:
"Felinies" was very much intentional

YakumoChen commented on code/game/objects/items.dm:
```diff
@@ -337,6 +346,82 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+ */
+/obj/item/proc/warning_label(var/list/readout = "")
+	var/list/crimes = list("Assaults", "Third Degree Murders", "Robberies", "Terrorist Attacks", "Different Felonies", "Felinies", "Counts of Tax Evasion", "Mutinies")
+	var/list/victims = list("Human", "Moth", "Felinid", "Lizard", "Particularly Resilient Slime", "Syndicate Agent", "Clown", "Mortal Foe", "Innocent Bystander")
```
Assistant
Rabid Pack of 5 Corgis
Wizard

cacogen replied:
This would look better if it was lowercase

ATH1909 commented:
okay but what if the item wasn't made by NT

StoneMason commented:
> okay but what if the item wasn't made by NT

Surely Nanotrasen would have databanks on most things in space, considering they’re one of the leading powers or whatever.

Pumpkinoe commented:
I feel like marking guns as 'DANGEROUS' only creates confusion for new players.

Iamgoofball commented:
i like this, this is good, good PR :+1:

cacogen commented:
This is sort of fascinating. Soulful. If it gets pulled I'm going to take it upon myself to do a spelling and grammar pass on it though because that aspect of it is killing me.

ATH1909 commented:
I think that it'd be best to combine this with the sechud PR, as ashwalkers or whatever seeing holograms in their nest on lavaland without any equipment that could explain that being nearby would be kind of weird

Qustinnus commented:
this should be an element probably

Qustinnus commented on code/game/objects/items.dm:
```diff
@@ -337,6 +346,82 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+	if(!override_notes)
+		/// Make sure not to divide by 0 on accident
+		if(force > 0)
+			readout += "Our EXTENSIVE research has shown that it takes a mere <span class='warning'>[round((100 / force), 0.1)]</span> hits to down a(n) [victims[rand(1, victims.len)]]."
```
byond has things to replace a/an automatically im pretty sure

Mothblocks replied:
yeah

Qustinnus commented on code/game/objects/items.dm:
```diff
@@ -337,6 +346,82 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+		else
+			readout += "If you decide to throw this object instead, then nice job, fuckwit. It doesn't hurt at all."
+		if(armour_penetration > 0 || block_chance > 0)
+			readout += "This item has demonstrated a(n) <span class='warning'>[weapon_tag_convert(armour_penetration)]</span> ability to pierce armor, and <span class='warning'>[weapon_tag_convert(block_chance)]</span> blocking capabilities."
```
ditto

SpaceDragon00 commented:
> I think that it'd be best to combine this with the sechud PR, as ashwalkers or whatever seeing holograms in their nest on lavaland without any equipment that could explain that being nearby would be kind of weird

Ashwalkers are an edge case the NT big shots want to account for, given how easy it is for them to stumble upon Nanotransen property, should you not consider lavaland as such already

Also, if I'm associating this with any item, it will very much not be a sechud, because this is supposed to be helpful to all players. If anything, I'll associate it with a tablet/computer program, but I'll avoid that til absolutely necessary because it's just more work to make something less helpful

> Oh yeah, a couple of ideas I had was having the amount of crimes be relative to the item's overall strength, e.g. force + throwforce + block_chance + armour_piercing

Not a bad idea but I think that more dangerous weapons would be harder to get. I'd reckon that more crimes are committed with kitchen knives than matter-destroying rocks on a sword

Greniza commented:
I feel like there should be some mention of what the 'ion' mode does for the multi-phase energy gun.

Does Nanotrasen possess automatic remote-tagging capabilities, for crafted items such as spears, baseball bats, and improvised shotguns?

I think that there should be a difference between "dangerous items" and "highly dangerous items" for the concern of security, like the difference between a baseball bat and a shotgun.

Another fun idea? Make it so that a toolbox has a very slim chance to be considered a 'dangerous item' and labeled as such, or add one of these labels to the heirloom toolbox.

norill commented:
I like the general idea as it lowers the (enormous) entry barrier, but i'm not a fan of them being holo labels. it just doesn't make sense in too many cases. labels on the lungs i just took out of my patient? i think it would be better executed by having an item (or a program) you could use to scan things

itseasytosee commented:
There are two ways to go this in my mind. You either hook this up to an eyepiece or scanner, or make a much less detailed version being just how it looks, but we already kind of have that.

cacogen commented on code/game/objects/items.dm:
```diff
@@ -337,6 +346,82 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+/obj/item/proc/weapon_tag_convert(var/tag_val)
+	switch(tag_val)
+		if(0)
+			return "NONEXISTANT"
```
non-existent

cacogen commented:
Damage values usually aren't obvious based on how items look. Then there's things like armour piercing and block chance that are applied inconsistently across items that you can't guess at all.

If it has to be the functionality of an object (and that's arbitrary because other object stats are already visible on examine) it shouldn't be an analyser because nobody wants to go around scanning weapons to figure out which one to pick. It should be the result of just examining it.

>I think that it'd be best to combine this with the sechud PR, as ashwalkers or whatever seeing holograms in their nest on lavaland without any equipment that could explain that being nearby would be kind of weird

You could argue it's all Nanotrasen owned space anyway and whatever technology is doing this is active throughout it. But I still think it's a mistake to try to ground it in the setting instead of it just being information told to the player on examine like what already exists. I do like this PR because of the lore though.

SpaceDragon00 commented:
Alright so: I've got 2 options that I'm considering here for how to make this work if it wouldn't be okay as is (Tablet program or Trait, like Self-aware), and I'll see if I can't talk about it with a maintainer

cacogen commented:
A powergame trait seems like such a bad idea. A tablet program would not only involve waiting for the program to download it'd also involve holding something in your hand to scan weapons which is clunky as fuck when you just want to grab something good to attack something with. It should show up on examine. If it has to be the product of an item, it would make sense if it were a HUD.

YakumoChen commented:
it's not really a powergaming trait since if you powergame you already know what the best weapons are

YakumoChen commented:
>Oh yeah, a couple of ideas I had was having the amount of crimes be relative to the item's overall strength, e.g. force + throwforce + block_chance + armour_piercing and I guess in the case of ballistic weapons that of the projectiles of their caliber and with energy weapons that of the sum of their fire modes.

Random numbers is funnier. I like the idea of 84 terrorist attacks committed in Nanotrasen space using only a piece of cardboard or a floor tile

cacogen commented:
I don't know why you would think adding a free trait that essentially becomes mandatory to take because of the combat advantage it offers that only people who actually bother to modify their traits will end up knowing about is a good idea.

I don't know either why you would think an obviously randomly generated statistic was funny. Would it be funny if it wasn't? Probably not. It would just be more believable and therefore more worthwhile to include.

Fikou commented:
i feel like this went from way too gamey to way too ingame
why not just have items have
"it seems like it would have a good chance at blocking an attack"
or
"this one seems like its robust at penetrating armor"

cacogen commented:
I have a simpler one up: #58858. This one is a fascinating piece of lore though. Just needs the text improved.

Mothblocks commented:
I talked it over with some people I trust and I'm going to take this PR over #58858 for some key reasons.

1. While I think obfuscating info is a powerful design tool, the cat is so out of the bag that it's already back at the pound. That is to say, basically anyone that cares code-dives, so I think it's okay to provide this information in some form or another.
2. This is better than direct numbers for everything (the very first PR) since at least it makes *some* effort to stay within world.
3. While I definitely like the way the other PR obscures data, it still feels like people will just inevitably code-dive or just test directly.

I have my own nitpicks about this design, but I am going to hesitantly choose this one.

Mothblocks commented:
As for implementation, make this entire thing an element.

MrDoomBringer commented:
> "it seems like it would have a good chance at blocking an attack"

Because players want to figure out if one item is marginally better than another.

SpaceDragon00 commented:
> 
> 
> As for implementation, make this entire thing an [element](https://hackmd.io/@tgstation/SignalsComponentsElements). This is one of your first PRs, so it is likely going to be pretty confusing, but the core conceit is that this information doesn't really need to be directly coupled with items, but instead just a generic "label" element, or whatever. If you need help with this, ping me on Discord @Mothblocks.

Alright, I'll get to work on this and try and get a commit up soon!

itseasytosee commented:
I doubt it will change anyone's minds but this is too silly and needlessly sci-fi for me. If it was part of an item or quirk I would not mind but because it's gonna show up on every examination of every item with a force it's gonna end up being annoying for people who don't care to read it.

Qustinnus commented:
![image](https://user-images.githubusercontent.com/20987591/117065369-b65fe600-ad27-11eb-8d55-1ad2aadfdd2a.png)

can people please read the pr this is like the tenth time i've seen the same message complaining about something that isn't an issue.

itseasytosee commented:
![image](https://user-images.githubusercontent.com/55666666/117065653-6106e300-aced-11eb-88ef-5fc65b5746cd.png)

coiax commented on code/game/objects/items.dm:
```diff
@@ -337,6 +346,82 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+	readout = list ("") /// Readout is used to store the text block output to the user so it all can be sent in one message
+
+	/// Meaningless flavor text
+	readout += "<span class='warning'>WARNING:</span> This item has been marked as dangerous by the NT legal team because of its use in <span class='warning'>[rand(2,99)] [crimes[rand(1, crimes.len)]].</span>"
```
But the number of crimes (and type of crime) will change if you examine again.

It really should be fixed per item class. Doesn't have to be fixed per round, I guess.
(Ideally, you'd like, do something like temporarily seed the random number generator with
the item's typepath, and then these values would be deterministic, but that's awkward to
 do in DM).

norill replied:
i would say either remove it or rework it into something meaningful. adding a meaningless number to this thing sounds fun at first, but it goes against the point of this feature. it would only confuse players and they will have to code-dive only to see that it's meaningless.

cacogen replied:
If it were me I'd just have this number secretly be (I know I suggested this earlier) the sum of its stats. That way it has a useful gameplay purpose (indicates overall how powerful the item is) in a way that's intuitive (more uses = better weapon). I don't think the logic that the more powerful an item is the rarer and less likely it is to be used in a crime matters here.

coiax commented on code/game/objects/items.dm:
```diff
@@ -299,6 +304,10 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
 
+	/// Code for showing offensive statistics via a hyperlinked tag
+	if(force >= 5 || throwforce >= 5 || override_notes || offensive_notes) /// Only show this tag for items that could feasibly be weapons, shields, or those that have special notes
+		. += "<span class='notice'>It appears to have an ever-updating holographic <a href='?src=[REF(src)];examine=s'>warning label.</a></span>"
```
I don't really like new Topic() handles, I feel like there should be a better way. Whether there actually is one, I don't know.

Have you considered putting this stuff into "examine more" instead?

Qustinnus replied:
examine more has terrible UI and noone knows it exists.

FlamingLily commented:
Hi there, I had a thought that maybe (and this may be a bit out of scope for this PR, but) perhaps both PRs that do this thing could be introduced? The more detailed stats in this PR can be shown if the user has a sechud (or maybe even a sec skillchip?), whereas the more general style of stats such as #58858 can be shown to those without?

Also keep in mind I come from downstream and have no idea how the average TG round plays, so take my suggestions with a grain of salt, but I thought this might be an interesting suggestion.

NullFag commented:
> Hi there, I had a thought that maybe (and this may be a bit out of scope for this PR, but) perhaps both PRs that do this thing could be introduced? The more detailed stats in this PR can be shown if the user has a sechud (or maybe even a sec skillchip?), whereas the more general style of stats such as #58858 can be shown to those without?
> 
> Also keep in mind I come from downstream and have no idea how the average TG round plays, so take my suggestions with a grain of salt, but I thought this might be an interesting suggestion.

­> Doesn't play on TG
­> Suggests things for it anyway

MrDoomBringer commented:
>  The more detailed stats in this PR can be shown if the user has a sechud (or maybe even a sec skillchip?), whereas the more general style of stats such as #58858 can be shown to those without?

Having 'general' (vague) stats about weapons won't solve any problems - people will still codedive for the actual useful information. Detailed information should be available to as many players as possible, which IMO means it should be on either an analyzer, PDA function, or just straight-up examine. We don't want new players forced to play sec / steal from sec just to figure out if their potted plant does more damage than a toolbox. 

@NullFag 
> angry noises

That guy contributes more to his downstream in a month than you do in a year, calm down lmao

FlamingLily commented:
> Having 'general' (vague) stats about weapons won't solve any problems - people will still codedive for the actual useful information. Detailed information should be available to as many players as possible, which IMO means it should be on either an analyzer, PDA function, or just straight-up examine. We don't want new players forced to play sec / steal from sec just to figure out if their potted plant does more damage than a toolbox.

Yeah that's fair enough.

SpaceDragon00 committed:
Changes Functionality to Element, Various Additional Improvements

- Migrates all of the isolated weapon description functionality to an element, which is added to all objects upon initialization

- Updated the text with various spelling fixes, flavor text improvements, and additional checks to increase clarity and improve flow

- Added the word 'bluespace'

SpaceDragon00 committed:
Merge branch 'tgstation:master' into master

SpaceDragon00 committed:
Merge branch 'master' into wepdescripts

SpaceDragon00 committed:
Adding a single break

Added a single break to make things look nicer

SpaceDragon00 commented:
This really should not have taken as long as it did, but god decided to punish me for my hubris, giving me 2 ears infections (In different ears), and then a stomach illness from the pain meds I took for those ear infection. Despite the sickness and my horrific work ethic, I have gotten around to improving this. If the hyperlinked line is considered too intrusive, a simple fix is to change COMSIG_PARENT_EXAMINE to COMSIG_PARENT_EXAMINE_CLOSER, to hide it behind a double examine. I tried to implement as many of the suggested changes as I could, but I felt some were unnecessary (You might even see why in my comments)

SpaceDragon00 committed:
Fixing Errors My Dipshit Self Made With Lighters

When you accidentally merge into master and have to fix something manually in another branch only to have to fix it again because it's detected as a change.......

SpaceDragon00 committed:
Fixing a little switcheroo on a boolean checkl

Fixes a backwards variable check so that the intended words are actually shown

SpaceDragon00 committed:
Removing Var/ From a Proc

Oops!

SpaceDragon00 committed:
Nanotransen -> Nanotrasen

I thought it was Nanotransen for months

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+/datum/element/weapon_description/proc/warning_label(atom/source, mob/user, list/examine_texts)
+	SIGNAL_HANDLER
+
+	var/obj/item/item = source
```
Just change the argument type of `source` to `/obj/item`.

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+/datum/element/weapon_description/proc/build_label_text(obj/item/source)
+	var/list/readout = list ("") /// Readout is used to store the text block output to the user so it all can be sent in one message
+
+	/// Meaningless flavor text. The number of crimes is constantly changing because of the complex nanotransen legal system and the esoteric nature of time itself!
```
`///` is only for doc comments (like for procs). Everything else can be `//`.

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+ *  * source - The object whose stats are being examined
+ */
+/datum/element/weapon_description/proc/build_label_text(obj/item/source)
+	var/list/readout = list ("") /// Readout is used to store the text block output to the user so it all can be sent in one message
```
```suggestion
	var/list/readout = list("") /// Readout is used to store the text block output to the user so it all can be sent in one message
```

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+	RegisterSignal(target, COMSIG_PARENT_EXAMINE, .proc/warning_label)
+	RegisterSignal(target, COMSIG_TOPIC, .proc/topic_handler)
+
+/datum/element/weapon_description/Detach(datum/target)
```
If you want your detach to be called, you need to add `element_flags = ELEMENT_DETACH`.

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+		readout += "Be especially careful around this device, as it is loaded with <span class='warning'>[source.magazine.caliber]</span> rounds, which you can inspect for more information."
+	else
+		readout += "The warning attached to the magazine is missing..."
+	return readout.Join("\n")
```
When you move these over, just make these return the string directly. There's no point in making a 1 list string then joining it.

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+	else
+		/// Labels don't do well with extreme forces
+		readout += "The warning label was blown away..."
+	return readout.Join("\n")
```
Same here, this is only ever returning one string.

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,185 @@
+		readout += source.offensive_notes
+
+	/// Spamming 'if's to see if we need to get additional weapon notes
+	if(istype(source, /obj/item/gun/ballistic))
+		readout += add_notes_ballistic(source)
+	else if(istype(source, /obj/item/gun/energy))
+		readout += add_notes_energy(source)
+	else if(istype(source, /obj/item/ammo_casing/))
+		readout += add_notes_ammo(source)
```
And also, you would add a new proc path called `/proc/add_weapon_description()` which these subtypes would override in order to make the element themselves.

Mothblocks commented:
Closer!

SpaceDragon00 committed:
Usr -> User

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Removing a silly space

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Contributor Suggestions + Additional Improvements

Changed the element into a bespoke element that adds an additional proc based on type

Added an obj/item proc that is used and overwritten to add the appropriate element to each item type

Minor improvements like changing var types, removing unnecessary code and spacing, and removing additional slashes from comments

Moved weapon notes code that must be calculated to initialize rather than examine

Minor wording improvements

SpaceDragon00 committed:
Merge branch 'wepdescripts' of https://github.com/SpaceDragon00/tgstation into wepdescripts

SpaceDragon00 committed:
Fixing division by 0 problems

SpaceDragon00 commented:
Thanks again for your help during all this, mothblocks!

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,186 @@
+			readout += "Our extensive research found that you couldn't beat yourself to death with this if you tried."
+
+		if(source.throwforce > 0)
+			readout += "If you decide to throw this object instead, one will take <span class='warning'>[round((100 / source.throwforce), 0.1)]</span> hits before collapsing."
```
```suggestion
			readout += "If you decide to throw this object instead, one will take <span class='warning'>[round((100 / source.throwforce), 0.1)]</span> hit\s before collapsing."
```

Otherwise "1 hits before collapsing"

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,186 @@
+ */
+/datum/element/weapon_description/proc/add_notes_ballistic(obj/item/gun/ballistic/source)
+	if(source.magazine) // Make sure you have a magazine, thats where the warning is!
+		return "\nBe especially careful around this device, as it can be loaded with <span class='warning'>[source.magazine.caliber]</span> rounds, which you can inspect for more information."
```
```suggestion
		return "\nBe especially careful around this device, as it can be loaded with <span class='warning'>[source.magazine.caliber]</span> round\s, which you can inspect for more information."
```

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,186 @@
+	if(!source.ammo_type)
+		return
+	var/obj/projectile/exam_proj
+	readout += "Standard models of this dangerous tool have <span class='warning'>[source.ammo_type.len]</span> mode(s)"
```
```suggestion
	readout += "Standard models of this dangerous tool have <span class='warning'>[source.ammo_type.len]</span> mode\s"
```

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,186 @@
+	for(var/obj/item/ammo_casing/energy/for_ammo in source.ammo_type)
+		exam_proj = for_ammo.loaded_projectile
+		if(exam_proj.damage > 0) // Don't divide by 0!!!!!
+			readout += "<span class='warning'>[round(100 / exam_proj.damage, 0.1)]</span> shots on <span class='warning'>[for_ammo.select_name]</span> mode before collapsing from [exam_proj.damage_type == STAMINA ? "immense pain" : "their wounds"]."
```
```suggestion
			readout += "<span class='warning'>[round(100 / exam_proj.damage, 0.1)]</span> shot\s on <span class='warning'>[for_ammo.select_name]</span> mode before collapsing from [exam_proj.damage_type == STAMINA ? "immense pain" : "their wounds"]."
```

Mothblocks commented on code/datums/elements/weapon_description.dm:
```diff
@@ -0,0 +1,186 @@
+	if(source.loaded_projectile)
+		// No dividing by 0
+		if(source.loaded_projectile.damage > 0)
+			return "Most monkeys our legal team subjected to these rounds succumbed to their wounds after <span class='warning'>[round(100 / (source.loaded_projectile.damage * source.pellets), 0.1)]</span> point-blank discharges, taking <span class='warning'>[source.pellets]</span> shots per round"
```
```suggestion
			return "Most monkeys our legal team subjected to these rounds succumbed to their wounds after <span class='warning'>[round(100 / (source.loaded_projectile.damage * source.pellets), 0.1)]</span> point-blank discharges, taking <span class='warning'>[source.pellets]</span> shot\s per round"
```

"point-blank discharges" should also become "point-blank discharge" at 1, but there is no `\es` so you'll need to make a variable.

Mothblocks commented on code/game/objects/items/melee/misc.dm:
```diff
@@ -205,6 +205,12 @@
+	. = ..()
+	// Adding an extra break for the sake of presentation
+	if(stamina_damage != 0)
+		offensive_notes = "\nVarious interviewed security forces report being able to beat criminals into exhaustion with only <span class='warning'>[round(100 / stamina_damage, 0.1)] hit(s)!</span>"
```
```suggestion
		offensive_notes = "\nVarious interviewed security forces report being able to beat criminals into exhaustion with only <span class='warning'>[round(100 / stamina_damage, 0.1)] hit\s!</span>"
```

Mothblocks commented on code/game/objects/items/stunbaton.dm:
```diff
@@ -52,6 +52,9 @@
 	. = ..()
+	// Adding an extra break for the sake of presentation
+	if(stamina_loss_amt != 0)
+		offensive_notes = "\nVarious interviewed security forces report being able to beat criminals into exhaustion with only <span class='warning'>[round(100 / stamina_loss_amt, 0.1)] hit(s)!</span>"
```
```suggestion
		offensive_notes = "\nVarious interviewed security forces report being able to beat criminals into exhaustion with only <span class='warning'>[round(100 / stamina_loss_amt, 0.1)] hit\s!</span>"
```

Mothblocks commented on code/modules/projectiles/ammunition/_ammunition.dm:
```diff
@@ -46,6 +47,9 @@
 	update_appearance()
 
+/obj/item/ammo_casing/add_weapon_description()
+	AddElement(/datum/element/weapon_description, attached_proc = /datum/element/weapon_description/proc/add_notes_ammo)
```
This should be on the ammo_casing item itself, not coupled to the element.

Mothblocks replied:
Remember this from our topic about the money stuff?

Mothblocks commented on code/modules/projectiles/guns/ballistic.dm:
```diff
@@ -131,6 +131,9 @@
 	RegisterSignal(src, COMSIG_ITEM_RECHARGED, .proc/instant_reload)
 
+/obj/item/gun/ballistic/add_weapon_description()
+	AddElement(/datum/element/weapon_description, attached_proc = /datum/element/weapon_description/proc/add_notes_ballistic)
```
Same here--proc on the gun, not on weapon_description.

Mothblocks commented on code/modules/projectiles/guns/energy.dm:
```diff
@@ -46,6 +46,9 @@
 	RegisterSignal(src, COMSIG_ITEM_RECHARGED, .proc/instant_recharge)
 
+/obj/item/gun/energy/add_weapon_description()
+	AddElement(/datum/element/weapon_description, attached_proc = /datum/element/weapon_description/proc/add_notes_energy)
```
Same.

Mothblocks reviewed (changes requested):
Closer.

SpaceDragon00 committed:
Update code/game/objects/items/stunbaton.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Update code/datums/elements/weapon_description.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Update code/datums/elements/weapon_description.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Update code/datums/elements/weapon_description.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Update code/game/objects/items/melee/misc.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SpaceDragon00 committed:
Contributor Suggestions + Additional Improvements

Moved the attached procs to the weapon types themselves

Added various /s-es where they were necessary, and removed some where they were not

Improved clarity on base message, to specify that you are physically hitting someone with the object in it

Moved the add element below Destroy() because splitting Initialize() and Destroy() felt bad

Removed source from the add_notes procs since they were no longer necessary

Fixed an oopsie where stamina-critting ammo types were not displaying their shots to stamina crit because of an overzealous removal of var/list/readout

SpaceDragon00 committed:
Merge branch 'tgstation:master' into wepdescripts
